### ISSUE TITLE
Too many header files need to be included when using C++ api

### ISSUE THREAD

**Issue by nicklhy** [apache/incubator-mxnet#9763]
Hi everyone, I am working on a C++ project which needs to use mxnet for CNN inference. I set `USE_CPP_PACKAGE = 1` and compiled mxnet source code myself.

In my C++ program, I only include `cpp-package/include/mxnet-cpp/MxNetCpp.h` at first but got errors like below:
```
/path/to/incubator-mxnet/cpp-package/include/mxnet-cpp/base.h:31:25: fatal error: mxnet/c_api.h: No such file or directory
```

It is caused by [`cpp-package/include/mxnet-cpp/base.h`](https://github.com/apache/incubator-mxnet/blob/master/cpp-package/include/mxnet-cpp/base.h#L31) which includes `mxnet/c_api.h` and `nnvm/c_api.h` inside it.
```
#include "mxnet/c_api.h"
#include "nnvm/c_api.h"
```
Thus, I have to add a lot more dirs to avoid the similar errors:
```
INCLUDE_DIRECTORIES(
    /path/to/incubator-mxnet/cpp-package/include
    /path/to/incubator-mxnet/include
    /path/to/incubator-mxnet/nnvm/include
    /path/to/incubator-mxnet/dmlc-core/include
    )
```

Just want to ask if it is possible to make the mxnet C++ api easier to use?

For instance, the mxnet cpp package folder only contains:
* `include`: all header files necessary for C++ development;
* `lib`: all libraries necessary for C++ development (i.e. libmxnet.so);

**Comment by anirudh2290** [comment]
@marcoabreu should we have a different tag for CPP_PACKAGE? I have seen a bunch of issues for the CPP_package  and it would be nice if we could group them together with a label.

**Comment by marcoabreu** [comment]
Good idea, please mention me in the other ones so I can label them accordingly

**Comment by sandeep-krishnamurthy** [comment]
There is discussion started by MXNet community to address this issue. Please chime in to provide your inputs and possibly contribute. 

https://lists.apache.org/thread.html/b8e655ab756db258efb1a5483b1643dbdc8fc0bfc12ce1cb8f1c2837@%3Cdev.mxnet.apache.org%3E

**Comment by meanmee** [comment]
error:
../../include/mxnet-cpp/optimizer.hpp:37:26: fatal error: mxnet-cpp/op.h: No such file or directory

But I can't find any files named op.h in mxnet-cpp, is this a bug?

**Comment by wenyangchu** [comment]
I have the same issue. It happens both when I build with mxnet c++ examples or gluon-cv examples
in gluon-cv/scripts/deployment/cpp-inference 
It is a very frustrating experience .. I have compile the source with USE_CPP_PACKAGE=1
I am searching for one working example how to link headers and packages

**Comment by yuyijie1995** [comment]
@meanmee  I met the same problam like you .Have you solved it?

**Comment by yuyijie1995** [comment]
@nicklhy  Hi , I have the same problem like you, but I do not understand “Thus, I have to add a lot more dirs to avoid the similar errors:” . I mean where to add these infomation ,I still confused about it .

**Comment by nicklhy** [comment]
@yuyijie1995 you can check the make/cmake files in the examples, something like [this](https://github.com/apache/incubator-mxnet/blob/da5242b732de39ad47d8ecee582f261ba5935fa9/cpp-package/example/Makefile#L31).

**Comment by yuyijie1995** [comment]
Thanks a lot




| |
俞依杰
|
|
邮箱：13588280094@163.com
|

签名由 网易邮箱大师 定制

On 01/28/2019 11:30, nicklhy wrote:

@yuyijie1995 you can check the make/cmake files in the examples, something like this.

**Comment by dexception** [comment]
Facing same issue.

**Comment by abbas-hassani** [comment]
Dear experts,
I have all the problems above and I am not an expert so I don't know how to fix these problems. 
Could you please tell me what directories should be included in what file?
Where are they and how I can find them?
Thank you in advance,
Abbas
